### PR TITLE
fix(es/typescript): Treat const variable references as enum constants

### LINKED ISSUE #11715: TypeScript enum transform does not treat const string variables as compile-time constants
### Describe the bug

SWC does not treat a local `const` string variable used in a TypeScript enum initializer as a compile-time constant.

For the input below, TypeScript 5.9 emits a string enum shape (`E["A"] = "aThisIs"`), but SWC emits reverse-mapping style output (`E[E["A"] = foo] = "A"`), which changes runtime behavior and enum shape.

### Input code

```typescript
const foo = 'aThisIs'

enum E {
  A = foo,
  B = 'bThisIs'
}
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript"
    }
  }
}
```

### Link to the code that reproduces this issue

https://github.com/nodejs/amaro/issues/394

### SWC Info output

```text
$ npx -y -p @swc/cli@latest -p @swc/core@latest swc -V
@swc/cli: 0.8.0
@swc/core: 1.15.18
```

### Expected behavior

SWC should match TypeScript enum lowering semantics for this case (string enum, no reverse mapping for `A`), e.g.:

```js
"use strict";
const foo = 'aThisIs';
var E;
(function (E) {
    E["A"] = "aThisIs";
    E["B"] = "bThisIs";
})(E || (E = {}));
```

### Actual behavior

SWC currently emits:

```js
var foo = 'aThisIs';
var E = function(E) {
    E[E["A"] = foo] = "A";
    E["B"] = "bThisIs";
    return E;
}(E || {});
```

### Version

@swc/core 1.15.18 (@swc/cli 0.8.0)

### Additional context

- Original downstream report: https://github.com/nodejs/amaro/issues/394
- Related esbuild issue: https://github.com/evanw/esbuild/issues/4387

### PR DESCRIPTION
**Pull request by baltasarblanco**
**Description:**

`EnumValueComputer::compute_rec` gates its identifier arm on `ctxt == unresolved_ctxt`, which only matches references to sibling enum members. A reference to a real `const` binding is a resolved binding, so it fell through to the `Opaque` catch-all.

Two consequences. The reported one is emit shape:

```ts
const foo = "aThisIs";
enum E { A = foo, B = "bThisIs" }
```

```js
// before — three properties, because the reverse mapping assigns twice
E[E["A"] = foo] = "A";
// after, and what tsc emits
E["A"] = "aThisIs";
```

The second is not in the issue and is worse — numeric members silently become `undefined`:

```ts
const foo = 1;
enum E { A = foo, B, C }
```

```
before -> { A: 1, "1": "A", "undefined": "C" }
after  -> { A: 1, B: 2, C: 3, "1": "A", "2": "B", "3": "C" }
```

`TsEnumRecordValue::inc()` returns `Void` for an opaque value, so every following auto-incremented member collapses.

This restores an existing code path rather than adding a rule: `compute_rec` already implements the constant folding from microsoft/TypeScript#50528 for literals, parens, unary and binary operators, template literals and enum member references. Const-variable references are the one production that was never wired in.

**What qualifies as a constant**

TypeScript decides this from the syntactic form alone, with no type resolution — confirmed as intentional and required for third-party transpilers in microsoft/TypeScript#63275, and described in evanw/esbuild#4387. A binding qualifies when it is a simple `const` with no type annotation, whose initializer is itself a constant expression, transitively — including ambient ones: `declare const x = 1` folds, since ambient contexts only allow initializers that are literals. Members of `declare enum` with constant initializers fold the same way — with the usual auto-increment inside an ambient `const enum`, none in a plain ambient enum, and runtime reads of the ambient object left untouched, all matching `tsc`. Two rules fall out of that and both are covered by fixtures:

- Type syntax removes constness. `const foo = "s" as string`, `const foo: "s" = "s"`, `foo satisfies T`, `<T>foo`, `foo!` and `as const` all stay reverse-mapped, and so does `enum E { A = foo as string }` — an assertion on the reference is enough.
- Within the same scope, a binding declared after the enum is not visible to it. The collection pass runs in source order, so a forward reference is simply not in the map yet — and it would be a TDZ error at runtime anyway. Across a function boundary `tsc` is more permissive; that case is listed under known limitations.

**Verified against `tsc` 5.9.3**

Every fixture case was diffed against `transpileModule` output. Folded: plain `const`, transitive chains, `"a" + "b"`, template literals, numeric bindings, and `const` bindings in function or namespace scope. Not folded: type annotations and assertions, `let`, `var`, annotated or uninitialized `declare const`, ambient enum members without initializers, destructured bindings, and forward references.

**Known limitations**

- Namespace member access (`N.foo`, `N.M.foo`) still produces a reverse mapping, and an auto-incremented member following it collapses to `undefined`. `compute_member` bails on any non-`Ident` object. Additive and independent; opened as #12102.
- An enum nested in a function-like body does not see outer `const` bindings declared later in the file. `tsc` folds those, since the body does not run at the declaration point: `function f() { enum E { A = x, B } } const x = 1` gives `A = 1`, `B = 2`. Matching it needs deferred evaluation of nested enums, which is a different rule from the syntactic one implemented here. Left as-is — it only ever under-folds, which is the behavior on `main` today.
- An ambient enum member referencing a `const` declared later in the file stays opaque — the collection pass runs in source order, while `tsc` folds it since ambient declarations are erased. Same deferred-evaluation family as the function-boundary case above.
- Cross-file constants are not folded. `tsc`'s own `transpileModule` does not fold them either — only `createProgram` does — so matching single-file semantics looks like the right contract here.
- Where `tsc` needs the checker, this stays conservative. `const foo: string = "s"` makes `tsc` emit a string enum without folding the value; a syntactic transpiler cannot know the type, so the member stays reverse-mapped.

`enum E { A = "s" as any }` still diverges from `tsc`, which reverse-maps it. That behavior predates this PR — it was introduced by #11769 for #11761 — so it is out of scope here; reported as #12150. `ts_enum_with_type_assertion` is unchanged.

**Tests:** six fixtures under `tests/fixture/`, plus two cases added to `ts_enum_is_mutable_true`. Each fails without the change — verified by copying them into a worktree checked out at `main`. Full crate suite green (203 fixture tests, 5039 identity), plus `swc --test tsc` (4579), `swc_ts_fast_strip` (4458), `swc --test projects` (889, including the `issues-11xxx/11761` fixture from #11769) and `swc --test exec` (451). One tsc conformance reference moved: in `constEnum2.ts`, `g = CONST` now resolves to a constant, so the member is inlined rather than emitted — the same treatment `d = 10` already gets in that same `const enum`. The members that call `Math.random()` stay opaque and are still emitted. Also checked with `RUSTFLAGS="--cfg swc_ast_unknown"` since this touches a `match` over `Expr`.

**Mutable enums**

Under `tsEnumIsMutable`, `enter_expr_for_inline_enum` deliberately leaves reads of non-const enums as runtime reads. The collection pass mirrors that guard: `const x = D.A` only resolves when `D` is a `const enum`. This diverges from `tsc`, which folds it — but matching `tsc` here would contradict the option's own contract, and would emit `const x = D.A` while using the folded value for the enum member that reads it. Both sides are covered by `ts_enum_is_mutable_true`: a const from a mutable enum member stays a runtime read, one from a `const enum` member still folds.

**Performance**

The first version of the collection pass evaluated every `const` initializer in the program. `compute` takes the expression by value, so each one was cloned, and in real TypeScript most `const` initializers — arrow functions, object literals, calls — hit the `Opaque` catch-all immediately, so the clone was discarded. That showed up as a CodSpeed regression on the TypeScript benchmarks.

Checking the outermost form before cloning removes it: `es/transform/baseline/common_typescript` goes from 59.6 us back to 53.3 us, against 53.3 us on `main`. Measured with `cargo bench` on both. No test output changes.

**BREAKING CHANGE:** None.

**Related issue (if exists):**
- Closes: #11715

### CODE REVIEW

**Comment by baltasarblanco** [comment]
Opened #12102 for the namespace member access limitation noted above.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 1]
```diff
@@ -536,6 +539,34 @@ impl Visit for SemanticAnalyzer {
...
+            let Pat::Ident(BindingIdent { id, type_ann: None }) = &decl.name else {
+                continue;
+            };
+            let Some(init) = &decl.init else { continue };
+
+            let empty = TsEnumRecord::default();
+            let value = EnumValueComputer {
+                enum_id: &id.to_id(),
+                unresolved_ctxt: self.unresolved_ctxt,
+                record: &empty,
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs]: Fixed in `d547a9a`. The collection pass now evaluates const initializers against the accumulated enum record instead of an empty one, so `const x = E.A` resolves. Verified against `tsc` 5.9.3; fixture `issue-11715-enum-member-init`, which fails on `main`.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 2]
```diff
@@ -536,6 +539,34 @@ impl Visit for SemanticAnalyzer {
...
+                const_vars: &self.info.const_vars,
+            }
+            .compute(init.clone(), EvalCtx::CONST_INIT);
+
+            if value.is_const() {
+                self.info.const_vars.insert(id.to_id(), value);
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs:636]: Confirmed against `tsc` 5.9.3, including the following member collapsing to `undefined`. Out of scope here: `compute_member` bails on any non-`Ident` object, which is an independent change. Opened as #12102.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 3]
```diff
@@ -566,6 +597,7 @@ impl Visit for SemanticAnalyzer {
                 &id.to_id(),
                 &default_init,
                 &self.info.enum_record,
+                &self.info.const_vars,
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs:674]: Confirmed against `tsc` 5.9.3. Matching it needs deferred evaluation of enums nested in function bodies, which is a different rule from the syntactic one this PR implements — top-level forward references must keep their current behavior, since `tsc` does not fold those either and they are a TDZ error at runtime. Left as-is; it only ever under-folds, which is what `main` does today. Noted under known limitations.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 4]
```diff
@@ -536,6 +539,34 @@ impl Visit for SemanticAnalyzer {
         }
     }
 
+    fn visit_var_decl(&mut self, node: &VarDecl) {
+        node.visit_children_with(self);
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs]: Fixed in `d547a9a`. Each declarator is recorded before the next one is traversed. Verified against `tsc` 5.9.3; fixture `issue-11715-declarator-order`, which fails on `main`.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 5]
```diff
@@ -536,6 +539,39 @@ impl Visit for SemanticAnalyzer {
...
+            let Some(init) = &decl.init else { continue };
+
+            if !EnumValueComputer::can_fold_shape(init) {
+                continue;
+            }
+
+            let value = EnumValueComputer {
+                enum_id: &id.to_id(),
+                unresolved_ctxt: self.unresolved_ctxt,
+                record: &self.info.enum_record,
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs:627]: Good catch — this was a real regression, fixed in `8eeebc5`. Confirmed by diffing against `main`: `enum D { A = 1 } (D as any).A = 5; const x = D.A; enum E { A = x }` emitted `E["A"] = 1` on this branch versus `E[E["A"] = x] = "A"` on `main`. The emitted code contradicted itself — `const x = D.A` was preserved while the enum member used the folded value.

The option is now threaded into the collection pass and mirrors the guard in `transform.rs`: with mutable enums enabled, only `const enum` members resolve through a member expression. Both sides are covered in `ts_enum_is_mutable_true`.

Separately, `enum F { A = D.A }` (no const in between) still folds under `tsEnumIsMutable` — that path goes through `transform_ts_enum_member`, is unchanged by this PR and behaves the same on `main`. I'll report it separately.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 6]
```diff
@@ -536,6 +543,40 @@ impl Visit for SemanticAnalyzer {
...
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs:627]: Both confirmed against `tsc` 5.9.3 and fixed in `6130967`. The original claim in the description — that `declare const` never folds — was wrong as written: it held only for the annotated or uninitialized forms. With a literal initializer (the only kind ambient contexts allow), `tsc` folds.

Ambient enum members with constant initializers go to a separate record that only the evaluator consults, so they fold inside enum and const initializers while runtime reads of the ambient object stay untouched — `console.log(A.X)` is not rewritten, matching `tsc`. Two edges pinned by the fixture: no auto-increment across ambient members (a member without an initializer stays opaque), and sibling references inside the ambient enum (`Y = X + 1`) fold. Fixture `issue-11715-ambient`, which fails on `main`.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 7]
```diff
@@ -536,6 +543,40 @@ impl Visit for SemanticAnalyzer {
         }
     }
 
+    fn visit_var_decl(&mut self, node: &VarDecl) {
+        let track = !self.skip_transform_info && !node.declare && node.kind == VarDeclKind::Const;
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs]: Confirmed against `tsc` 5.9.3 and fixed in `6130967`. `declare const x = 1` now populates `const_vars` like any other const; the declaration itself is still erased. The description's "non-ambient" restriction was based on measuring only the annotated form (`declare const x: number`), which indeed does not fold — corrected there too. Fixture `issue-11715-ambient`, which fails on `main`.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 8]
```diff
@@ -536,10 +580,46 @@ impl Visit for SemanticAnalyzer {
...
+            }
+        }
+    }
+
     fn visit_ts_enum_decl(&mut self, node: &TsEnumDecl) {
         node.visit_children_with(self);
 
         if self.skip_transform_info {
+            self.record_ambient_enum(node);
             return;
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs:646]: Confirmed and fixed in `ca3679b`. Ambient `const enum` ids are now registered in `const_enum` before the ambient path returns, so under `tsEnumIsMutable` their members resolve like any other `const enum` — only non-const enums stay mutable. Covered in `ts_enum_is_mutable_true`: `declare const enum M { A = 7 }` through a const folds to `7` / `8` with the option enabled. Loose runtime reads of the ambient object are still not rewritten: the inliner keys on `enum_record`, which ambient enums never enter.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 9]
```diff
@@ -285,6 +295,40 @@ impl SemanticAnalyzer {
...
+impl SemanticAnalyzer {
+    /// Ambient enums are erased from the output, but `tsc` still treats their
+    /// members with constant initializers as constant enum expressions inside
+    /// enum and const initializers. There is no auto-increment in ambient
+    /// enums: a member without an initializer stays opaque. Kept out of
+    /// `enum_record` so the inliner never rewrites runtime reads of the
+    /// ambient object.
+    fn record_ambient_enum(&mut self, node: &TsEnumDecl) {
+        for member in &node.members {
+            let Some(init) = &member.init else { continue };
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs]: Confirmed against `tsc` 5.9.3 and fixed in `ca3679b` — my earlier no-auto-increment rule held only for plain ambient enums. In an ambient `const enum` every member is constant, so the usual sequence applies, including re-anchoring after an explicit initializer: `declare const enum A { X, Y = 5, Z }` gives `X = 0`, `Z = 6`, matching `tsc`. Pinned in `issue-11715-ambient`. Plain ambient enums keep the measured behavior: a member without an initializer stays opaque.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 10]
```diff
@@ -285,6 +296,50 @@ impl SemanticAnalyzer {
...
+        }
+
+        let mut default_init: TsEnumRecordValue = 0.0.into();
+
+        for member in &node.members {
+            let value = match (&member.init, node.is_const) {
+                (Some(init), _) => EnumValueComputer {
+                    enum_id: &node.id.to_id(),
+                    unresolved_ctxt: self.unresolved_ctxt,
+                    record: &self.info.ambient_enum_record,
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs]: Confirmed against `tsc` 5.9.3 and fixed in `9976d67ebe`. `record_ambient_enum` now evaluates against the normal enum record, with the ambient fallback extended to the sibling arm so bare references inside the ambient enum (`Y = X + 1`) keep resolving; const bindings resolve too, and the split reproduces enum merging (ambient half folds in initializers, runtime reads stay untouched). The mutable-enum guard applies here as well. Pinned in `issue-11715-ambient`, which fails on `main`.

File: crates/swc_ecma_transforms_typescript/src/ts_enum.rs [hunk 11]
```diff
@@ -278,17 +388,29 @@ impl EnumValueComputer<'_> {
...
+            member_name,
+        };
+
         self.record
-            .get(&TsEnumRecordKey {
-                enum_id: ident.to_id(),
-                member_name,
-            })
+            .get(&key)
+            .or_else(|| self.ambient_record.get(&key))
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/ts_enum.rs]: Confirmed and fixed in `b7a2e5a9e2`. The ambient lookup is now gated on `allow_const_var`, which `compute_rec` already clears when evaluation crosses type syntax — so an ambient member reached through an assertion stays opaque, directly, through a const binding, and in the sibling arm (`declare enum A { X = 1, Y = X as number }` leaves `A.Y` opaque). Same on `tsc` 5.9.3, so the divergence is not specific to 6.0.3. All pinned in `issue-11715-ambient`.

The concrete counterpart is unchanged and still diverges on `main` as well: `enum E { A = C.X as number }` folds where `tsc` keeps the runtime access. That predates this PR (#11769) and is tracked in #12150 — narrowing it here would revert that PR's fixture.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 12]
```diff
@@ -285,6 +296,50 @@ impl SemanticAnalyzer {
...
+            let value = match (&member.init, node.is_const) {
+                (Some(init), _) => EnumValueComputer {
+                    enum_id: &node.id.to_id(),
+                    unresolved_ctxt: self.unresolved_ctxt,
+                    record: &self.info.enum_record,
+                    const_vars: &self.info.const_vars,
+                    const_enum_only: self.ts_enum_is_mutable.then_some(&self.info.const_enum),
+                    ambient_record: &self.info.ambient_enum_record,
+                }
+                .compute(init.clone(), EvalCtx::MEMBER),
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs]: Confirmed and fixed in `7bffbb1`. Ambient member initializers now evaluate under the const-initializer rule (`ts_is_opaque`), so type syntax anywhere in the initializer makes the member opaque instead of being stripped — `declare enum A { X = 1 as number }` stays a runtime read downstream, matching `tsc` 5.9.3. Pinned in `issue-11715-ambient`.

File: crates/swc_ecma_transforms_typescript/src/ts_enum.rs [hunk 13]
```diff
@@ -117,20 +130,98 @@ pub(crate) fn static_enum_member_name(property: &MemberProp) -> Option<Wtf8Atom>
...
+    /// under-fold; it can never produce a wrong value.
+    pub fn can_fold_shape(expr: &Expr) -> bool {
+        matches!(
+            expr,
+            Expr::Lit(..)
+                | Expr::Ident(..)
+                | Expr::Paren(..)
+                | Expr::Unary(..)
+                | Expr::Bin(..)
+                | Expr::Member(..)
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/ts_enum.rs:200]: Confirmed against `tsc` 5.9.3 — it does fold `E?.A`. This is an under-fold (identical emit to today's `main`), and this PR's collection scope is closed, so it's reported separately as #12152 rather than growing the diff further.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 14]
```diff
@@ -285,6 +296,50 @@ impl SemanticAnalyzer {
     }
 }
 
+impl SemanticAnalyzer {
+    /// Ambient enums are erased from the output, but `tsc` still treats their
+    /// members with constant initializers as constant enum expressions inside
+    /// enum and const initializers. In an ambient `const enum` every member is
+    /// constant, so the usual auto-increment applies; in a plain ambient enum
+    /// a member without an initializer stays opaque. Kept out of `enum_record`
+    /// so the inliner never rewrites runtime reads of the ambient object.
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs:307]: Measured the opposite on the reference this PR declares: `tsc` 5.9.3 `transpileModule` emits `const x = A.X;` unchanged — it does not inline ambient const enum members. Inlining them is checker behavior (`createProgram`), which single-file transpilation deliberately doesn't match, per the contract stated in the description ("only `createProgram` does — so matching single-file semantics looks like the right contract"). SWC's emit here is also byte-identical to `main`'s, so nothing regressed. Leaving as is.

File: crates/swc_ecma_transforms_typescript/src/semantic.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs]: Confirmed — measured the same, and it's already the last entry under known limitations in the description: the collection pass runs in source order, so an ambient member referencing a later `const` stays opaque, while `tsc` folds it since ambient declarations are erased. Matching it needs deferred evaluation, the same family as the function-boundary case listed above it. Under-fold only — identical emit to today's `main` — and this PR's collection scope is closed.

**Review by kdy1 — COMMENTED** [review]
I added a test case

File: crates/swc_ecma_transforms_typescript/src/semantic.rs [hunk 15]
```diff
@@ -267,8 +275,11 @@ impl SemanticAnalyzer {
                     enum_id,
                     unresolved_ctxt,
                     record,
+                    const_vars,
+                    const_enum_only: None,
+                    ambient_record,
```
[the remark this answers is not in the document]
Comment (baltasarblanco) [reply crates/swc_ecma_transforms_typescript/src/semantic.rs:281]: Confirmed and fixed in `7385335acb` — kdy1's test pins it and passes unchanged. The ambient fallback (member lookup and sibling arm) is now gated on the same const-enum rule the option applies elsewhere; the primary record is untouched, since concrete enum reads under the option predate this PR (#12151).

**Comment by baltasarblanco** [comment]
Thanks for adding the test — fixed in `7385335acb` and it passes unchanged. The ambient fallback is now gated on the same const-enum rule the option applies elsewhere; concrete enum reads under the option are untouched, since that behavior predates this PR (#12151, which has a contributor working on it).

**Review by kdy1 — COMMENTED** [review]
Image

This PR regressed the perf. Can you investigate?

**Comment by baltasarblanco** [comment]
Investigated. Absolute walltime per commit, same machine and session (`es/transform/baseline/common_typescript`):

| commit | time |
|---|---|
| `3c4f404` (PR base) | 52.3 µs |
| `f6f9b27` | 53.2 µs |
| `8eeebc5` | 53.1 / 52.7 µs |
| `7bffbb1` (ambient folding) | 54.3 / 53.8 µs |
| `7385335` (HEAD) | 54.4 µs |
| `e876e80` (`main` today) | 51.6 µs |

The regression is real, reproducible (two A/B pairs), ~+2%, and comes entirely from `7bffbb1`. Part of the −2% CodSpeed shows is `main` having improved ~1.5% while this branch sits on the older base.

Profiled both sides (`perf`, dwarf call graphs): no single hot function — the benchmark input has no `enum` and no `declare`, so the ambient machinery never executes on it, and the delta is spread across allocation and the widened collection pass (±0.3pp each on `malloc`/clones/AST walk). I tried two targeted gates on the collection pass; one breaks the ambient fixtures, the other keeps all 205 tests green but doesn't recover the time — so this looks like the structural cost of the extra pass state rather than a fixable hot path.

My read: the ambient folding fixes real corruption (members collapsing to `undefined`, pinned in `issue-11715-ambient`, fails on `main`), which is worth 2% on this microbench — so unless you prefer otherwise, I'll leave it as is. If you'd rather keep the baseline flat, say the word and I'll move the ambient support to a follow-up: a revert commit on top (no history rewrite — your mutable-ambient test pins preservation and keeps passing either way) and reopen it once the cost is understood.

**Review by kdy1 — APPROVED** [review]
thanks!
